File: app/src/main/java/com/chaitupenjudcoder/recyclerviews/BucksOverviewRecycler.java
package com.chaitupenjudcoder.recyclerviews;

import android.content.Context;
import androidx.databinding.DataBindingUtil;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.chaitupenjudcoder.buckstrack.R;
import com.chaitupenjudcoder.buckstrack.databinding.ItemCategoryTotalListBinding;
import com.chaitupenjudcoder.datapojos.CategoriesAmount;

import java.util.ArrayList;

public class BucksOverviewRecycler extends RecyclerView.Adapter<BucksOverviewRecycler.BucksOverviewAdapter> {

    private Context ctx; ArrayList<CategoriesAmount> catAmounts;
    public BucksOverviewRecycler(Context ctx, ArrayList<CategoriesAmount> catAmounts) {
        this.ctx = ctx;
        this.catAmounts = catAmounts;
    }

    @NonNull
    @Override
    public BucksOverviewAdapter onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        LayoutInflater inflater = LayoutInflater.from(viewGroup.getContext());
        ItemCategoryTotalListBinding items = DataBindingUtil.inflate(inflater, R.layout.item_category_total_list, viewGroup, false);
        return new BucksOverviewRecycler.BucksOverviewAdapter(items);
    }

    @Override
    public void onBindViewHolder(@NonNull BucksOverviewAdapter bucksOverviewAdapter, int i) {
        CategoriesAmount cAmount = catAmounts.get(i);
        bucksOverviewAdapter.itemViewBinding.tvCategory.setText(cAmount.getCategoryName());
        bucksOverviewAdapter.itemViewBinding.tvAmount.setText(cAmount.getTotalAmount());
        int percentage = Math.round(cAmount.getPercentage());
        bucksOverviewAdapter.itemViewBinding.pbCategory.setProgress(percentage);
    }

    @Override
    public int getItemCount() {
        return catAmounts.size();
    }

    class BucksOverviewAdapter extends RecyclerView.ViewHolder {
        ItemCategoryTotalListBinding itemViewBinding;

        public BucksOverviewAdapter(@NonNull ItemCategoryTotalListBinding itemViewBinding) {
            super(itemViewBinding.getRoot());
            this.itemViewBinding = itemViewBinding;
        }
    }
}

File: app/src/main/java/com/chaitupenjudcoder/recyclerviews/BucksOverviewRecycler.kt
package com.chaitupenjudcoder.recyclerviews

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.chaitupenjudcoder.buckstrack.R
import com.chaitupenjudcoder.buckstrack.databinding.ItemCategoryTotalListBinding
import com.chaitupenjudcoder.datapojos.CategoriesAmount
import java.util.*
import kotlin.math.roundToInt

class BucksOverviewRecycler(var catAmounts: ArrayList<CategoriesAmount>) :
    RecyclerView.Adapter<BucksOverviewRecycler.BucksOverviewAdapter>() {

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int) = BucksOverviewAdapter(
        DataBindingUtil.inflate(
            LayoutInflater.from(viewGroup.context),
            R.layout.item_category_total_list, viewGroup, false
        )
    )

    override fun onBindViewHolder(holder: BucksOverviewAdapter, position: Int) {
        holder.bind(catAmounts[position])
    }

    override fun getItemCount(): Int = catAmounts.size


    inner class BucksOverviewAdapter(private val itemViewBinding: ItemCategoryTotalListBinding) :
        RecyclerView.ViewHolder(itemViewBinding.root) {

        fun bind(catAmount: CategoriesAmount) {
            val (categoryName, totalAmount, percentage) = catAmount

            itemViewBinding.apply {
                tvCategory.text = categoryName
                tvAmount.text = totalAmount
                pbCategory.progress = percentage.roundToInt()
            }
        }
    }
}